source/HardwareSerial.cpp

```diff
@@ -31,276 +61,559 @@ HardwareSerial::~HardwareSerial()
-};
-
 HardwareSerial::operator bool(void)
 {
-    if (_comHandle != NULL && _comHandle != INVALID_HANDLE_VALUE)
-    {
-        return true;
-    }
-    return false;
+    return true;
 }
 
 int HardwareSerial::available(void)
 {
-    if ((_storageCount - _storageIndex) <= 0)
+    std::deque<IBuffer^>::const_iterator iter;
+    int bytesAvailable = 0;
+
+    EnterCriticalSection(&m_readBufferListLock);
+    iter = m_readBufferList.begin();
+    while (iter != m_readBufferList.end())
     {
-        peek();
+        bytesAvailable += (*iter)->Length;
+        ++iter;
     }
-    return _storageCount - _storageIndex;
+    LeaveCriticalSection(&m_readBufferListLock);
+
+    return bytesAvailable;
 }
 
 void HardwareSerial::setTimeout(unsigned long timeout)
 {
-    _timeout = timeout;
-
-    // Sets WIN32 READ/WRITE timeouts
-    COMMTIMEOUTS CommTimeouts;
-    GetCommTimeouts(_comHandle, &CommTimeouts);
+    m_timeout = timeout;
+}
 
-    CommTimeouts.ReadIntervalTimeout = 0;
-    CommTimeouts.ReadTotalTimeoutConstant = _timeout;
-    CommTimeouts.ReadTotalTimeoutMultiplier = 0;
+void HardwareSerial::begin(unsigned long baud, uint8_t config)
+{
+    HRESULT hr = S_OK;
 
-    CommTimeouts.WriteTotalTimeoutConstant = _timeout;
-    CommTimeouts.WriteTotalTimeoutMultiplier = 0;
+    hr = g_pins.verifyPinFunction(m_txPinNo, FUNC_SER, BoardPinsClass::LOCK_FUNCTION);
+    if (FAILED(hr))
+    {
+        ThrowError(hr, "An error occurred configuring pin %d as for Serial TX use.", m_txPinNo);
+    }
 
-    if (!SetCommTimeouts(_comHandle, &CommTimeouts))
+    hr = g_pins.verifyPinFunction(m_rxPinNo, FUNC_SER, BoardPinsClass::LOCK_FUNCTION);
+    if (FAILED(hr))
     {
-#ifdef _DEBUG
-        Log("Error %d when setting Com timeouts: ", GetLastError());
-        LogLastError();
-#endif
+        ThrowError(hr, "An error occurred configuring pin %d for Serial RX use.", m_rxPinNo);
     }
-}
 
-void HardwareSerial::begin(unsigned long baud, uint8_t config)
-{
-    DCB dcb = dcbArray[config];
+    // Set up objects needed to cancel read threads when .end() is called.
+    m_cancellationTokenSource = new Concurrency::cancellation_token_source();
+    Concurrency::cancellation_token cToken(m_cancellationTokenSource->get_token());
+
+    m_serialConfig = (SerialConfigs)config;
+    m_baudRate = baud;
 
-    if ( _comPortName == L"\\\\.\\COM1" )
+    m_currentReadNumber = 0;
+    m_queuedReadCount = 0;
+    m_peekByte = NO_PEEK_BYTE;
+    m_readThreadCount = 0;
+
+    // Create event we use to determine when the UART open operation has completed.  This is used,
+    // rather than .wait() becaue .begin() may be called from the main thread when an object in a 
+    // library is constructed--and neither .wait() nor .get() can be called from the main thread.
+    std::shared_ptr<Concurrency::event> findCompleted = std::make_shared<Concurrency::event>();
```

Comment:
If this is just a consistency thing, I would like to defer this until there are changes that require retesting.